### PR TITLE
support for checking op's inputs grad semantic

### PR DESCRIPTION
**Pull request by lxd-cumt**
### PR types

New features
### PR changes

Others
### Description

Pcard-66975
Support for checking whether inputs of OP have grad semantic

Taking MultiplyOp as an example, the OpInputInfo generated before and after supporting input grad semantic is as follows:
```c++
OpInfoTuple MultiplyOp::GetOpInfo() {
  std::vector<paddle::dialect::OpInputInfo> inputs = { paddle::dialect::OpInputInfo("x", "paddle::dialect::DenseTensorType", false, false, false), paddle::dialect::OpInputInfo("y", "paddle::dialect::DenseTensorType", false, false, false) };
  std::vector<paddle::dialect::OpAttributeInfo> attributes = {  };
  ...
  return std::make_tuple(inputs, attributes, outputs, run_time_info, "multiply");
}
```
```c++
OpInfoTuple MultiplyOp::GetOpInfo() {
  std::vector<paddle::dialect::OpInputInfo> inputs = { paddle::dialect::OpInputInfo("x", "paddle::dialect::DenseTensorType", false, false, false, true), paddle::dialect::OpInputInfo("y", "paddle::dialect::DenseTensorType", false, false, false, true) };
  std::vector<paddle::dialect::OpAttributeInfo> attributes = {  };
  ...
  return std::make_tuple(inputs, attributes, outputs, run_time_info, "multiply");
}
```

### CODE REVIEW

File: paddle/fluid/ir/dialect/op_generator/op_gen.py [hunk 1]
```diff
@@ -704,6 +721,69 @@ def to_pascal_case(s):
         return "".join([word.capitalize() for word in words]) + ""
 
 
+def OpInputGradSemanticCheck(op_info, op_info_items):
```
Comment (Charles-hit) [inline paddle/fluid/ir/dialect/op_generator/op_gen.py:724]: python 函数命名用小写 + 下划线，跟上面保持一致，函数名是不是叫`get_input_grad_semantic`更合适一点?

File: paddle/fluid/ir/dialect/op_generator/op_gen.py [hunk 2]
```diff
@@ -704,6 +721,69 @@ def to_pascal_case(s):
...
+                    input_grad_semantic_list.append("true")
+                else:
+                    input_grad_semantic_list.append("false")
+        else:
+            input_grad_semantic_list = ["false" for i in range(num_inputs)]
+
+    return input_grad_semantic_list
+
+
+def OpMutableAttributeGradSemanticCheck(op_info, op_info_items):
```
Comment (Charles-hit) [inline paddle/fluid/ir/dialect/op_generator/op_gen.py:758]: 同上 ，函数名用`get_mutable_attribute_grad_semantic`？

File: paddle/fluid/ir/dialect/op_generator/op_gen.py [hunk 3]
```diff
@@ -704,6 +721,69 @@ def to_pascal_case(s):
...
+        for bwd_output in bwd_output_list:
+            bwd_output_list_new.append(bwd_output[:-5])
+
+        for i in range(len(fwd_mutable_attribute_list)):
+            if fwd_mutable_attribute_list[i] in bwd_output_list_new:
+                mutable_attribute_grad_semantic_list.append("true")
+            else:
+                mutable_attribute_grad_semantic_list.append("false")
+
+    return mutable_attribute_grad_semantic_list
```
Comment (Charles-hit) [inline paddle/fluid/ir/dialect/op_generator/op_gen.py:784]: 这个变量叫`mutable_attribute_grad_semantics`更合适一点？ 尽量不要在变量命名中加类型

File: paddle/fluid/ir/dialect/op_generator/op_gen.py [hunk 4]
```diff
@@ -704,6 +721,69 @@ def to_pascal_case(s):
...
+            ), "Configuration of forward op and backward op is not match."
+            for i in range(num_inputs):
+                if bwd_fwd_input_list[i] in bwd_output_list_new:
+                    input_grad_semantic_list.append("true")
+                else:
+                    input_grad_semantic_list.append("false")
+        else:
+            input_grad_semantic_list = ["false" for i in range(num_inputs)]
+
+    return input_grad_semantic_list
```
Comment (Charles-hit) [inline paddle/fluid/ir/dialect/op_generator/op_gen.py:755]: 变量名是不是可以叫`input_grad_semantics`？

File: test/ir/test_op_input_grad_semantic.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+import unittest
+
+import paddle
+from paddle import ir
+
+paddle.enable_static()
+
+
+def get_ir_program_0():
```
Comment (Charles-hit) [inline test/ir/test_op_input_grad_semantic.py:23]: 这个函数是不是可以叫`get_gather_program_new_ir`?

File: test/ir/test_op_input_grad_semantic.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            shape=[3, 4], dtype='float32', value=2.0
+        )
+        index = paddle.tensor.fill_constant(shape=[1], dtype='int32', value=1.0)
+        axis = paddle.tensor.fill_constant(shape=[1], dtype='int32', value=2.0)
+        out = paddle.gather(x, index, axis)
+    newir_program = ir.translate_to_new_ir(main_program.desc)
+    return newir_program
+
+
+def get_ir_program_1():
```
Comment (Charles-hit) [inline test/ir/test_op_input_grad_semantic.py:39]: `get_multiply_program_new_ir`

File: test/ir/test_op_input_grad_semantic.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        y = paddle.tensor.fill_constant(
+            shape=[3, 4], dtype='float32', value=3.0
+        )
+        out = paddle.multiply(x, y)
+    newir_program = ir.translate_to_new_ir(main_program.desc)
+    return newir_program
+
+
+class TestOpInputGradSemantic(unittest.TestCase):
+    def test_gatherop_input_grad_semantic(self):
```
Comment (Charles-hit) [inline test/ir/test_op_input_grad_semantic.py:57]: 函数名可以去掉op或者用下划线隔开

File: test/ir/test_op_input_grad_semantic.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        newir_program = get_ir_program_0()
+        op = newir_program.block().ops[-1]
```
Comment (Charles-hit) [inline test/ir/test_op_input_grad_semantic.py:59]: op->gather_op

File: test/ir/test_op_input_grad_semantic.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        self.assertEqual(op.get_input_grad_semantics(), [True, False, False])
+
+    def test_multiplyop_input_grad_semantic(self):
+        newir_program = get_ir_program_1()
+        op = newir_program.block().ops[-1]
```
Comment (Charles-hit) [inline test/ir/test_op_input_grad_semantic.py:64]: op->multiply_op

**Review by zhangbo9674 — APPROVED** [review]

File: paddle/fluid/ir/dialect/op_generator/op_gen.py (hunk 1 above)
Comment (lxd-cumt) [reply paddle/fluid/ir/dialect/op_generator/op_gen.py:724]: done, thx

File: test/ir/test_op_input_grad_semantic.py (hunk 9 above)
Comment (lxd-cumt) [reply test/ir/test_op_input_grad_semantic.py:64]: done, thx

File: paddle/fluid/ir/dialect/op_generator/op_gen.py (hunk 2 above)
Comment (lxd-cumt) [reply paddle/fluid/ir/dialect/op_generator/op_gen.py:758]: done, thx

File: paddle/fluid/ir/dialect/op_generator/op_gen.py (hunk 3 above)
Comment (lxd-cumt) [reply paddle/fluid/ir/dialect/op_generator/op_gen.py:784]: done, thx

File: paddle/fluid/ir/dialect/op_generator/op_gen.py (hunk 4 above)
Comment (lxd-cumt) [reply paddle/fluid/ir/dialect/op_generator/op_gen.py:755]: done, thx

File: test/ir/test_op_input_grad_semantic.py (hunk 5 above)
Comment (lxd-cumt) [reply test/ir/test_op_input_grad_semantic.py:23]: done, thx

File: test/ir/test_op_input_grad_semantic.py (hunk 6 above)
Comment (lxd-cumt) [reply test/ir/test_op_input_grad_semantic.py:39]: done, thx

File: test/ir/test_op_input_grad_semantic.py (hunk 7 above)
Comment (lxd-cumt) [reply test/ir/test_op_input_grad_semantic.py:57]: done, thx

File: test/ir/test_op_input_grad_semantic.py (hunk 8 above)
Comment (lxd-cumt) [reply test/ir/test_op_input_grad_semantic.py:59]: done, thx

**Pull request merged**